File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.java
class Solution {
    Set<Character> vowelSet = Set.of('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U');

    private void swap(char[] chars, int left, int right) {
        char tmp = chars[left];
        chars[left] = chars[right];
        chars[right] = tmp;
    }

    public String reverseVowels(String s) {
        char[] chars = s.toCharArray();
        int n = chars.length;
        int left = 0;
        int right = n - 1;

        while (left < right) {
            if (vowelSet.contains(chars[left]) && vowelSet.contains(chars[right])) {
                swap(chars, left, right);
                left++;
                right--;
            } else if (vowelSet.contains(chars[left])) {
                right--;
            } else if (vowelSet.contains(chars[right])) {
                left++;
            } else {
                left++;
                right--;
            }
        }

        return new String(chars);
    }
}

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.kt
class Solution {
    private inline fun Char.isVowel(): Boolean {
        return this in "aeiouAEIOU"
    }

    private inline fun CharArray.swap(i: Int, j: Int) {
        this[i] = this[j].also { this[j] = this[i] }
    }

    fun reverseVowels(s: String): String {
        val chars = s.toCharArray()
        var left = 0
        var right = chars.lastIndex

        while (left < right) {
            val leftIsVowel = chars[left].isVowel()
            val rightIsVowel = chars[right].isVowel()

            when {
                leftIsVowel && rightIsVowel -> {
                    chars.swap(left, right)
                    left++
                    right--
                }
                leftIsVowel -> right--
                rightIsVowel -> left++
                else -> {
                    left++
                    right--
                }
            }
        }

        return String(chars)
    }
}